src/Compilers/CSharp/Portable/Symbols/ReferenceManager.cs

```diff
@@ -1104,6 +1102,17 @@ private static ImmutableArray<AssemblyIdentity> GetReferencedAssemblies(CSharpCo
 
                     Debug.Assert(sourceReferencedAssemblies.Length == sourceReferencedAssemblySymbols.Length);
 
+                    // Pre-calculate size to ensure this code only requires a single array allocation.
+                    var builderSize = modules.Sum(static (module, index) =>
+                    {
+                        if (index == 0)
+                            return module.GetReferencedAssemblySymbols().Count(static identity => !identity.IsLinked);
+                        else
+                            return module.GetReferencedAssemblies().Length;
+                    });
+
+                    var result = ImmutableArray.CreateBuilder<AssemblyIdentity>(builderSize);
```

Comment:
this will still leak the builder.  if you use ArrayBuilder, along with ToImmutableAndClear, then the array will be moved out, and the builder will put back into the pool.

Reply:
This will leak the builder, but using `ArrayBuilder<>` will cause it to leak the actual pooled array instance, which is (in most cases) larger than the builder itself.

Reply:
In theory we could modify `ArrayBuilder<T>.GetInstance(int)` to do something like this:

```cs
public static ArrayBuilder<T> GetInstance(int capacity)
{
    var builder = GetInstance();
    if (builder.Capacity < capacity)
    {
        if (builder.Capacity < 4)
        {
            // Keep this builder, but discard its backing array which is too small
            builder.EnsureCapacity(capacity);
        }
        else
        {
            // Return this builder to the pool, and create a new one with an appropriately-sized backing array
            builder.Free();
            builder = new ArrayBuilder<T>(s_poolInstance, capacity);
        }
    }

    return builder;
}
```

Reply:
It's kind of unfortunate that this allocates no matter what if the capacity of the retrieved builder isn't large enough. 

There are probably big reasons not to do it this way, but what about if we just had the arraybuilder keep an empty builder that it could use in the GetInstance(capacity) count? I tried it out in https://github.com/dotnet/roslyn/pull/71800 and it seems to work, but I haven't tested extensively. It feels like this approach gives users of ArrayBuilder.GetInstance a chance to get back an already allocated builder (that hasn't already allocated the underlying array)